subiquity/server/ad_joiner.py

```diff
@@ -0,0 +1,117 @@
+# Copyright 2023 Canonical, Ltd.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Affero General Public License as
+# published by the Free Software Foundation, either version 3 of the
+# License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Affero General Public License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+import asyncio
+import logging
+from socket import gethostname
+from subiquitycore.utils import arun_command
+from subiquity.server.curtin import run_curtin_command
+from subiquity.common.types import (
+    ADConnectionInfo,
+    AdJoinResult,
+)
+
+log = logging.getLogger('subiquity.server.ad_joiner')
+
+
+class AdJoinStrategy():
+    realm = "/usr/sbin/realm"
+    pam = "/usr/sbin/pam-auth-update"
+
+    def __init__(self, app):
+        self.app = app
+
+    async def do_join(self, info: ADConnectionInfo, hostname: str, context) \
+            -> AdJoinResult:
+        """ This method changes the hostname and perform a real AD join, thus
+            should only run in a live session. """
+        result = AdJoinResult.JOIN_ERROR
+        hostname_current = gethostname()
+        # Set hostname for AD to determine FQDN (no FQDN option in realm join,
+        # only adcli, which only understands the live system, but not chroot)
+        cp = await arun_command(['hostname', hostname])
+        if cp.returncode:
+            log.info("Failed to set live session hostname for adcli")
+            return result
+
+        root_dir = self.app.root
+        cp = await run_curtin_command(
+            self.app, context, "in-target", "-t", root_dir,
+            "--", self.realm, "join", "--install", root_dir, "--user",
+            info.admin_name, "--computer-name", hostname, "--unattended",
+            info.domain_name, private_mounts=True, input=info.password,
+            timeout=60)
+
+        if not cp.returncode:
+            # Enable pam_mkhomedir
+            cp = await run_curtin_command(self.app, context, "in-target",
+                                          "-t", root_dir, "--",
+                                          self.pam, "--package",
+                                          "--enable", "mkhomedir",
+                                          private_mounts=True)
+
+            if cp.returncode:
+                result = AdJoinResult.PAM_ERROR
+            else:
+                result = AdJoinResult.OK
+
+        # Restoring the live session hostname.
+        cp = await arun_command(['hostname', hostname_current])
```

Comment:
It's a bit unusual to set a hostname temporarily, but I'll take it for the moment as a requirement.

It would be nice to structure the set hostname / restore hostname in something that is more certain to get restored.  For instance a context manager could be created which sets the hostname, you run your in-targets, then it is responsible for restoring the hostname.

The subiquity codebase has several context managers, journald_subscriptions is a short example that I think is pretty clear.